### PR TITLE
fix: extract VRF to a plugin #248

### PR DESCRIPTION
**Pull request by ppedziwiatr**
`main`:
image

`ppe/vrf-plugin`:
image
😎

### CODE REVIEW

File: src/contract/HandlerBasedContract.ts [hunk 1]
```diff
@@ -688,7 +689,13 @@ export class HandlerBasedContract<State> implements Contract<State> {
     dummyTx.sortKey = await this._sorter.createSortKey(dummyTx.block.id, dummyTx.id, dummyTx.block.height, true);
     dummyTx.strict = strict;
     if (vrf) {
-      dummyTx.vrf = generateMockVrf(dummyTx.sortKey, arweave);
+      Arweave.utils;
```
Comment (Tadeuchi) [inline src/contract/HandlerBasedContract.ts:692]: you can remove this line and then the `import Arweave`

Comment (ppedziwiatr) [reply src/contract/HandlerBasedContract.ts:692]: shit, wtf is this

File: package.json [hunk 2]
```diff
@@ -95,6 +93,7 @@
     "warp-wasm-metering": "1.0.1"
   },
   "devDependencies": {
+    "@idena/vrf-js": "^1.0.1",
```
Comment (asiaziola) [inline package.json:96]: ? what for

File: package.json [hunk 3]
```diff
@@ -103,6 +102,7 @@
     "arlocal": "1.1.42",
     "cheerio": "^1.0.0-rc.10",
     "colors": "^1.4.0",
+    "elliptic": "^6.5.4",
```
Comment (asiaziola) [inline package.json:105]: what for?

File: src/core/modules/impl/ArweaveGatewayInteractionsLoader.ts [hunk 4]
```diff
@@ -116,15 +117,17 @@ export class ArweaveGatewayInteractionsLoader implements InteractionsLoader {
...
-            return t.name == WARP_TAGS.REQUEST_VRF && t.value === 'true';
-          })
-        ) {
-          interaction.vrf = generateMockVrf(interaction.sortKey, this.arweave);
+        if (hasVrfTag(interaction)) {
+          if (vrfPlugin) {
+            interaction.vrf = vrfPlugin.process().generateMockVrf(interaction.sortKey);
+          } else {
+            this.logger.warn('Cannot generate mock vrf for interaction - no "warp-contracts-plugin-vrf" attached!');
+          }
```
Comment (asiaziola) [inline src/core/modules/impl/ArweaveGatewayInteractionsLoader.ts:130]: hmm, seems tripled. can't we generate a custom function for the vrf generation?
edit: it's even quadrupled

File: src/core/modules/impl/ArweaveGatewayInteractionsLoader.ts [hunk 5]
```diff
@@ -142,5 +145,12 @@ export class ArweaveGatewayInteractionsLoader implements InteractionsLoader {
...
     this.arweaveTransactionQuery = new ArweaveGQLTxsFetcher(warp);
+    this._warp = warp;
   }
 }
+
+export function hasVrfTag(interaction: GQLNodeInterface) {
+  return interaction.tags.some((t) => {
+    return t.name == WARP_TAGS.REQUEST_VRF && t.value === 'true';
+  });
+}
```
Comment (asiaziola) [inline src/core/modules/impl/ArweaveGatewayInteractionsLoader.ts:156]: hmm, rest of the functions like this one (`isInteractWrite`, `isEvmSigned`) are in the TagsParser class. Maybe let's stick to the convention. Seems more like a generic function but I leave it to you to decide.

File: package.json (hunk 2 above)
Comment (ppedziwiatr) [reply package.json:96]: for tests :-) https://github.com/warp-contracts/warp/pull/401/files#diff-fc8f73adefb2d436efca49ad1f9867b9ba3e40610db08a6c574ccbb92565ca74R22

File: package.json (hunk 3 above)
Comment (ppedziwiatr) [reply package.json:105]: same shit

File: src/core/modules/impl/ArweaveGatewayInteractionsLoader.ts (hunk 5 above)
Comment (ppedziwiatr) [reply src/core/modules/impl/ArweaveGatewayInteractionsLoader.ts:156]: done!

**Pull request merged**